### PR TITLE
fix: retry v17 keeper portfolio on LaserStream fast path

### LINKED ISSUE #253: [MEDIUM] LaserStream fast-path discovery does not retry v17 keeper portfolio provisioning
## Summary

When `KEEPER_USE_LASERSTREAM=true`, `CrankService.discover()` can return through the LaserStream fast-path without retrying v17 keeper portfolio provisioning for already-discovered markets whose `keeperPortfolio` is still `null`.

This can leave a v17 market discovered but uncrankable until the next full rediscovery, because the v17 crank path skips when `state.keeperPortfolio` is not provisioned.

## Root Cause

The normal rediscovery path retries keeper portfolio provisioning when a market is already known but `state.keeperPortfolio` is still missing.

However, the LaserStream fast-path returns the existing in-memory markets directly:

```ts
return Array.from(this.markets.values()).map((s) => s.market);
```

Before returning, it updates owner-verified information from the LaserStream account loader cache, but it does not retry `ensureKeeperPortfolio()` for v17 markets whose `keeperPortfolio` remains `null`.

Later, the v17 crank path skips the market when `state.keeperPortfolio` is missing:

```ts
if (!state.keeperPortfolio) {
  logger.debug("Skipping crank — keeper portfolio not provisioned yet", { slabAddress });
  return false;
}
```

So a transient provisioning failure can leave the market present in memory but skipped by crank execution until a full rediscovery occurs.

## Impact

This can temporarily strand v17 markets in a discovered-but-uncrankable state.

Potential consequences:

1. A v17 market remains known to the keeper but cannot be cranked because `keeperPortfolio` is still `null`.
2. LaserStream fast-path discovery does not retry provisioning, so the missing portfolio state can persist.
3. The market may only recover on the next full rediscovery interval.
4. During that window, keeper execution for the affected market can be delayed.

This does not directly steal funds, but it can reduce keeper liveness for affected v17 markets.

Severity: Medium.

## Proof of Concept

Add this test to `tests/services/crank.test.ts` inside the `describe('discover', () => { ... })` block, before `describe('crankMarket', () => { ... })`:

```ts
it('PoC: LaserStream fast-path does not retry v17 keeper portfolio provisioning when keeperPortfolio is null', async () => {
  const prevLaserStream = process.env.KEEPER_USE_LASERSTREAM;
  process.env.KEEPER_USE_LASERSTREAM = 'true';

  try {
    const slabAddress = new PublicKey('11111111111111111111111111111111');

    const mockCache = {
      getOwnerVerified: vi.fn(() => null),
    };

    const mockAccountLoader = {
      getCache: vi.fn(() => mockCache),
      getStats: vi.fn(() => ({ lastSlot: 123 })),
      getProgramId: vi.fn(() => new PublicKey('11111111111111111111111111111111')),
    };

    const service = new CrankService(mockOracleService, undefined, mockAccountLoader as any);

    const mockMarket = {
      slabAddress,
      programId: new PublicKey('11111111111111111111111111111111'),
      config: {
        collateralMint: new PublicKey('11111111111111111111111111111111'),
        oracleAuthority: PublicKey.default,
        indexFeedId: { toBytes: () => new Uint8Array(32) },
      },
      params: { maintenanceMarginBps: 500n },
      header: { admin: PublicKey.default },
    };

    // Simulate a v17 market already discovered earlier, but portfolio provisioning failed.
    // In this state, crankMarket() skips because keeperPortfolio is still null.
    (service as any).markets.set(slabAddress.toBase58(), {
      market: mockMarket,
      lastCrankTime: 0,
      successCount: 0,
      failureCount: 0,
      consecutiveFailures: 0,
      isActive: true,
      missingDiscoveryCount: 0,
      keeperPortfolio: null,
    });

    // Keep discover() inside the LaserStream fast-path instead of full rediscovery.
    (service as any)._lastFullRediscoverTime = Date.now();

    vi.mocked(shared.getConnection).mockClear();

    await service.discover();

    // If the fast-path retried ensureKeeperPortfolio(), it would call getConnection()
    // before entering provisionKeeperPortfolio().
    await Promise.resolve();
    await Promise.resolve();

    expect(shared.getConnection).toHaveBeenCalled();

    service.stop();
  } finally {
    if (prevLaserStream === undefined) {
      delete process.env.KEEPER_USE_LASERSTREAM;
    } else {
      process.env.KEEPER_USE_LASERSTREAM = prevLaserStream;
    }
  }
});
```

Run:

```bash
pnpm exec vitest run tests/services/crank.test.ts -t "PoC: LaserStream fast-path"
```

Current result:

```text
FAIL tests/services/crank.test.ts > CrankService > discover > PoC: LaserStream fast-path does not retry v17 keeper portfolio provisioning when keeperPortfolio is null

AssertionError: expected "vi.fn()" to be called at least once

expect(shared.getConnection).toHaveBeenCalled()
```

This shows that `ensureKeeperPortfolio()` is not retried in the LaserStream fast-path for an already-discovered v17 market with `keeperPortfolio: null`.

## Expected Behavior

LaserStream fast-path discovery should retry keeper portfolio provisioning for v17 markets whose `keeperPortfolio` is still `null`, just like the normal rediscovery path does.

## Actual Behavior

LaserStream fast-path discovery returns existing markets without retrying keeper portfolio provisioning.

As a result, a v17 market can remain discovered but skipped by crank execution because `state.keeperPortfolio` remains `null`.

## Suggested Fix

Before returning from the LaserStream fast-path, retry keeper portfolio provisioning for any active market whose `keeperPortfolio` is still missing.

Suggested shape:

```ts
for (const [, state] of this.markets) {
  const key = state.market.slabAddress.toBase58();

  // existing LaserStream owner-verified refresh logic...

  if (!state.keeperPortfolio) {
    this.ensureKeeperPortfolio(key, state.market);
  }
}
```

The existing provisioning in-flight guard should prevent duplicate concurrent provisioning attempts, so retrying from the fast-path should be safe.

## Difference from #238 / #250 / #251

#238 reported that normal rediscovery did not retry v17 keeper portfolio provisioning after a transient failure. #250 fixed that by extracting `ensureKeeperPortfolio()` and calling it on first discovery and again on normal rediscovery when `keeperPortfolio` is still null. #251 tracks regression coverage for that normal rediscovery path.

This issue is narrower and separate: when `KEEPER_USE_LASERSTREAM=true`, `discover()` can return through the LaserStream fast-path before the normal rediscovery branch runs. That fast-path refreshes cached market data and returns existing markets, but it does not call `ensureKeeperPortfolio()` for already-discovered v17 markets whose `keeperPortfolio` remains null.

So the normal rediscovery retry is fixed, but the LaserStream fast-path still misses the retry.

### PR DESCRIPTION
**Pull request by Bayyan16**
## Summary

Fixes #253.

This PR makes the LaserStream fast-path discovery retry v17 keeper portfolio provisioning for already-discovered markets whose `keeperPortfolio` is still `null`.

Without this retry, a transient provisioning failure could leave a v17 market present in memory but uncrankable until the next full rediscovery, because the v17 crank path skips markets without a keeper portfolio.

## Why

The normal rediscovery path already retries keeper portfolio provisioning when `keeperPortfolio` is still missing.

The LaserStream fast-path returns existing in-memory markets before the normal rediscovery branch runs. It refreshes cached account data, but previously did not retry `ensureKeeperPortfolio()` for markets whose keeper portfolio was still missing.

This keeps the LaserStream fast-path consistent with the normal rediscovery behavior.

## Changes

* Retry `ensureKeeperPortfolio()` inside the LaserStream fast-path when `state.keeperPortfolio` is still `null`.
* Keep the existing owner-verified cache refresh logic unchanged.
* Add a regression test covering an already-discovered v17 market with `keeperPortfolio: null` while `KEEPER_USE_LASERSTREAM=true`.

## How to test

1. Run the targeted regression test:

```bash
pnpm exec vitest run tests/services/crank.test.ts -t "LaserStream fast-path retries"
```

2. Run the full crank service test file:

```bash
pnpm exec vitest run tests/services/crank.test.ts
```

3. Run the full test suite:

```bash
pnpm test
```

## Checklist

* [x] Linked issue included: Fixes #253
* [x] Regression test added
* [x] Targeted regression test passes
* [x] `tests/services/crank.test.ts` passes
* [x] Full test suite passes
* [x] No unrelated files changed

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved keeper portfolio provisioning reliability by ensuring proper initialization during accelerated market discovery operations.

* **Tests**
  * Added test coverage for keeper portfolio provisioning edge cases.

### CODE REVIEW

File: src/services/crank.ts [hunk 1]
```diff
@@ -921,6 +921,15 @@ export class CrankService {
...
             }
           }
+
+          // v17 markets need a keeper-owned portfolio before PermissionlessCrank
+          // can run. Normal rediscovery retries this when keeperPortfolio is
+          // still null; keep the LaserStream fast path consistent so a transient
+          // provisioning failure does not leave an already-known market
+          // discovered-but-uncrankable until the next full rediscovery.
+          if (!state.keeperPortfolio) {
+            this.ensureKeeperPortfolio(key, state.market);
```
[the remark this answers is not in the document]
Comment (Bayyan16) [reply src/services/crank.ts]: Fixed in the latest push.

The LaserStream fast-path retry is now gated to v17 markets only, using a header-safe v17 guard before calling ensureKeeperPortfolio(). I also tightened the regression test so it verifies the portfolio lookup path via getProgramAccounts and includes a legacy market with keeperPortfolio: null to ensure the retry is not attempted for non-v17 markets.

**Comment by Bayyan16** [comment]
Updated in the latest push.

The regression test now asserts the provisioning lookup path directly instead of only checking shared.getConnection(). It also includes a legacy market with keeperPortfolio: null to avoid false positives and ensure the fast-path retry only applies to v17 markets.
